fix pulsatrix charger - leak, race, error handling, stats

sauttefk opened the pull request:
Fix context leak in wsReader
Fix race condition in webSocketWrite
Add missing error handling for write operations
Add connection statistics

sauttefk committed:
Fix context leak in wsReader
Fix race condition in webSocketWrite
Add missing error handling for write operations

sauttefk committed:
typos

sauttefk committed:
add connection stats

sauttefk committed:
Merge branch 'master' into fix/pulsatrix-leak-race

sauttefk committed:
fix linter findings

sauttefk committed:
Merge branch 'evcc-io:master' into fix/pulsatrix-leak-race

sauttefk commented on charger/pulsatrix.go:
```diff
@@ -81,9 +95,10 @@ func NewPulsatrixFromConfig(other map[string]interface{}) (api.Charger, error) {
-		uri:  fmt.Sprintf("ws://%s/api/ws", hostname),
-		data: util.NewMonitor[pulsatrixData](15 * time.Second),
+		log:   util.NewLogger("pulsatrix"),
+		uri:   fmt.Sprintf("ws://%s/api/ws", hostname),
```
supply equipment charging controller does not support wss.
comparable to modbus communication with charging controllers

sauttefk committed:
fix sourcery.ai finding

sauttefk commented on charger/pulsatrix.go:
```diff
@@ -81,9 +95,10 @@
-		uri:  fmt.Sprintf("ws://%s/api/ws", hostname),
-		data: util.NewMonitor[pulsatrixData](15 * time.Second),
+		log:   util.NewLogger("pulsatrix"),
+		uri:   fmt.Sprintf("ws://%s/api/ws", hostname),
```
supply equipment charge controller does not support WSS connections
comparable to ModBus-TCP connections

sauttefk committed:
be less noisy
only warn after 3 failures

sauttefk committed:
Merge branch 'evcc-io:master' into fix/pulsatrix-leak-race

sauttefk committed:
cut-n-paste error fixed

andig commented:
I would appreciate if we could remove all the additional statistics and logging again once you're find that it's working as expected. Please also use the provided `ctx` to manage the lifetime of the background go routines.

sauttefk committed:
added module description
constants for easier configuration
faster mutex
general code cleanup

sauttefk committed:
according to @andig requests
- remove connection statistics
- using the provided ctx for lifetime and background routines

sauttefk committed:
Merge branch 'master' into fix/pulsatrix-leak-race

andig commented on charger/pulsatrix.go:
```diff
@@ -80,155 +110,291 @@ func NewPulsatrixFromConfig(other map[string]interface{}) (api.Charger, error) {
+	}
+
+	if err := backoff.Retry(operation, bo); err != nil {
+		c.log.ERROR.Printf("reconnect to pulsatrix SECC at %s failed: %v", c.hostname, err)
```
...an der Stelle brauchts dann Neustart?

sauttefk replied:
im prinzip ja.
wenn du möchtest kann ich nach diesen 15 minuten auch weiterhin versuchen neu zu verbinden.

sauttefk replied:
OK.
der reconnect wird jetzt unbegrenzt versucht, aber alle 15 minuten wird ein fehler geloggt.

andig commented:
Das diff ist schwer zu lesen weil du anscheinend die meisten Methodennamen geändert hast. Falls möglich wäre es hilfreich, auf ein *kleines* Diff zu schauen um es verstehen zu können. Lint/Clean Fehler sind auch noch vorhanden.

sauttefk commented:
@andig sorry for submitting the PR a bit early. After fixing the main issues, I had a few more ideas to make the implementation more robust.  
In the meantime, the integration has been running very stable, logging has been reduced to a minimum, and I’ve also incorporated your suggestion.  
I hope this makes the PR ready to be merged.

sauttefk committed:
fix whitespace and linefeed issues

sauttefk committed:
try to reconnect forever
log reconnect problem every 15 minutes

sauttefk committed:
remove unused contant

sauttefk committed:
Merge branch 'master' into fix/pulsatrix-leak-race

sauttefk commented:
@andig die Änderungen laufen seit 2 Wochen stabil und ohne Probleme. Ich denke das kann gemerged werden.

andig commented:
Viele der Änderungen, die Umbenennungen, aber z.B. auch https://github.com/evcc-io/evcc/pull/23281/files#diff-7481cf4f2ea1a56caad6ba56aa184ff8f4d9f2ba64b3c4ec5a47ca7e79993960L268-R446 machen das Diff größer, scheinen aber keine inhaltliche Änderung zu sein. Ich kann leider nicht bewerten, ob/was dieser PR ändert und ob er technisch gut aussieht. Wenn Du ihn rein haben willst müssen wir das ohne Review tun.
